sklearn/linear_model/huber.py

```diff
@@ -0,0 +1,349 @@
+        The iteration will stop when max{|proj g_i | i = 1, ..., n} <= pgtol
+        where pg_i is the i-th component of the projected gradient.
+
+    Attributes
+    ----------
+    coef_: array, shape (n_features,)
+        Features got by optimizing the Huber loss.
+
+    intercept_: float
+        Bias factor.
+
+    scale_ : float
+        The value by which |y - X'w - c| is scaled down by.
+
+    n_iter_: int
+        Number of iterations that fmin_l_bfgs_b has run for, if available.
+
+    References
+    ----------
+    .. [1] Peter J. Huber, Elvezio M. Ronchetti, Robust Statistics
+           Concomitant scale estimates, pg 172
+    .. [2] Art B. Owen (2006), A robust hybrid of lasso and ridge regression.
+           http://statweb.stanford.edu/~owen/reports/hhu.pdf
+    """
+
+    def __init__(self, epsilon=1.35, n_iter=100, alpha=0.0001,
+                 warm_start=False, fit_intercept=True, pgtol=1e-05):
+        self.epsilon = epsilon
+        self.n_iter = n_iter
+        self.alpha = alpha
+        self.warm_start = warm_start
+        self.fit_intercept = fit_intercept
+        self.pgtol = pgtol
+
+    def fit(self, X, y, sample_weight=None):
+        """Fit the model according to the given training data.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Training vector, where n_samples in the number of samples and
+            n_features is the number of features.
+
+        y : array-like, shape (n_samples,)
+            Target vector relative to X.
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+        X, y = check_X_y(
+            X, y, copy=False, accept_sparse=['csr'], dtype=np.float64, y_numeric=True)
+        if sample_weight is not None:
+            sample_weight = np.array(sample_weight)
+            check_consistent_length(X, y, sample_weight)
+        else:
+            check_consistent_length(X, y)
+
+        if self.epsilon < 1.0:
+            raise ValueError(
+                "epsilon should be greater than 1.0, got %f" % self.epsilon)
+        coef = getattr(self, 'coef_', None)
+        if not self.warm_start or (self.warm_start and coef is None):
+            if self.fit_intercept:
+                self.coef_ = np.zeros(X.shape[1] + 2)
+            else:
+                self.coef_ = np.zeros(X.shape[1] + 1)
+        elif self.warm_start:
+            self.coef_ = np.concatenate(
+                (self.coef_, [self.intercept_, self.scale_]))
+
+        # Sigma or the scale factor should be non-negative.
+        # Setting it to be zero might cause undefined bounds hence we set it
+        # to a value close to zero.
+        bounds = np.tile([-np.inf, np.inf], (self.coef_.shape[0], 1))
+        bounds[-1][0] = 1e-12
+
+        try:
+            self.coef_, f, self.dict_ = optimize.fmin_l_bfgs_b(
+                _huber_loss_and_gradient, self.coef_, approx_grad=True,
```

Comment:
TypeError is caused in old versions of scipy because of no maxiter argument. Will add a comment